src/components/App.tsx

```diff
@@ -87,24 +92,49 @@ class App extends React.Component<Props, MyState> {
           />
           <PrivateRoute
             path="/:id/edit"
-            render={(props: RouteComponentProps<{ id: string }>) => (
-              <HabitForm
-                {...props}
-                now={new Date()}
-                id={props.match.params.id}
-              />
-            )}
+            render={(props: RouteComponentProps<{ id: string }>) => {
+              const parsedId = +props.match.params.id;
+              const habit = !isNaN(parsedId)
+                ? this.props.userHabits[parsedId]
+                : null;
+              if (habit) {
+                return (
+                  <HabitForm
+                    {...props}
+                    now={new Date()}
+                    id={parsedId}
+                    habit={habit}
+                  />
+                );
+              } else {
+                return <NotFound />;
+              }
+            }}
           />
           <PrivateRoute
             path="/:id/show"
-            render={(props: RouteComponentProps<{ id: string }>) => (
-              <HabitCard id={props.match.params.id} collapsed={false} />
-            )}
+            render={(props: RouteComponentProps<{ id: string }>) => {
+              const parsedId = +props.match.params.id;
+              const habit = !isNaN(parsedId)
+                ? this.props.userHabits[parsedId]
+                : null;
+              if (habit) {
+                return (
+                  <HabitCard id={parsedId} habit={habit} collapsed={false} />
+                );
+              } else {
+                return <NotFound />;
+              }
+            }}
           />
         </Switch>
       </Router>
     );
-  }
+  };
 }
 
+const NotFound = () => {
+  return <span>Not Found</span>;
+};
+
```

Comment:
We should break this out into its own component (and ideally make it look a bit nicer over time, though this is fine for now).